testing/async/mixin.h

```diff
@@ -0,0 +1,645 @@
+#include <thrust/iterator/discard_iterator.h>
+
+#include <unittest/unittest.h>
+#include <unittest/util_async.h>
+
+#include <tuple>
+#include <type_traits>
+
+// This file contains a set of mix-in classes that define an algorithm
+// definition for use with test_policy_overloads<algo_def>. The algorithm
+// definition describes the details of a thrust::async algorithm invocation:
+//
+// - Input type and initialization
+// - Output type and initialization (supports in-place, too)
+// - Postfix arguments that define the algorithm's overload set
+// - Abstracted invocation of the async algorithm
+// - Abstracted invocation of a reference algorithm
+// - Validation of async vs. reference output
+// - A description string.
+//
+// This definition is used by test_policy_overloads to test each overload
+// against a reference while injecting a variety of execution policies. This
+// validates that each overload behaves correctly according to some reference.
+//
+// Since much of the algorithm definition is generic and may be reused in
+// multiple tests with slight changes, a mix-in system is used to simplify
+// the creation of algorithm definitions. The following namespace hierarchy is
+// used to organize these generic components:
+//
+// * testing::async::mixin::
+// ** ::input - Input types/values (device vectors, counting iterators, etc)
+// ** ::output - Output types/values (device vectors, inplace device vectors,
+//                                    discard iterators, etc)
+// ** ::postfix_args - Algorithm specific overload sets
+// ** ::invoke_reference - Algorithm specific reference invocation
+// ** ::invoke_async - Algorithm specific async algo invocation
+// ** ::compare_outputs - Compare output values.
+//
+// Each algorithm should define its own `mixins.h` header to declare algorithm
+// specific mixins (e.g. postfix_args, invoke_reference, and invoke_async)
+// in a testing::async::<algorithm_name>::mixins namespace structure.
+//
+// For example, the test.async.exclusive_scan.basic test uses the following
+// algorithm definition from mix-ins:
+//
+// ```
+//   #include <async/test_policy_overloads.h>
+//   #include <async/mixin.h>
+//   #include <async/exclusive_scan/mixin.h>
+//   template <typename input_value_type,
+//            typename output_value_type   = input_value_type,
+//            typename initial_value_type  = input_value_type,
+//            typename alternate_binary_op = thrust::maximum<>>
+//   struct basic_invoker
+//      : testing::async::mixin::input::device_vector<input_value_type>
+//      , testing::async::mixin::output::device_vector<output_value_type>
+//      , testing::async::exclusive_scan::mixin::postfix_args::
+//          all_overloads<initial_value_type, alternate_binary_op>
+//      , testing::async::exclusive_scan::mixin::invoke_reference::
+//          host_synchronous<input_value_type, output_value_type>
+//      , testing::async::exclusive_scan::mixin::invoke_async::basic
```

Comment:
We should call `basic` simple because `simple` is the name used in existing.

Reply:
Done.